core/ledger/kvledger/snapshot_mgmt_test.go

```diff
@@ -110,3 +117,198 @@ func TestSnapshotRequestBookKeeperErrorPaths(t *testing.T) {
+	defer cleanup()
+	provider := testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	defer provider.Close()
+
+	// create a ledger with genesis block
+	ledgerID := "testsnapshotrequests"
+	bg, gb := testutil.NewBlockGenerator(t, ledgerID, false)
+	gbHash := protoutil.BlockHeaderHash(gb.Header)
+	l, err := provider.Create(gb)
+	require.NoError(t, err)
+	defer l.Close()
+	kvledger := l.(*kvLedger)
+
+	// Test 1: submit a request at height 5 that is higher than block height and verify snapshot is generated when block height matches the request
+	require.NoError(t, l.SubmitSnapshotRequest(5))
+	require.Equal(t, uint64(5), kvledger.snapshotRequestBookkeeper.smallestRequestHeight)
+
+	// commit blocks until block height is 10
+	lastBlock := testutilCommitBlocks(t, l, bg, 10, gbHash)
+	// wait until snapshot height=5 is generated
+	snapshotExists := func() bool {
+		exists, err := kvledger.snapshotExists(5)
+		require.NoError(t, err)
+		return exists
+	}
+	require.Eventually(t, snapshotExists, time.Minute, 3*time.Second)
+
+	// Test 2: submit a request at height=0 so that generateSnapshot is called immediately
+	require.NoError(t, l.SubmitSnapshotRequest(0))
+	// wait until snapshot height=10 is generated
+	snapshotExists = func() bool {
+		exists, err := kvledger.snapshotExists(10)
+		require.NoError(t, err)
+		return exists
+	}
+	require.Eventually(t, snapshotExists, time.Minute, 3*time.Second)
+
+	// Test 3: submit multiple requests in parallel, verify PendingSnapshotRequest and smallest request height
+	for _, height := range []uint64{100, 15, 50, 20} {
+		go l.SubmitSnapshotRequest(height)
+	}
+	// wait until all requests are submitted and previous snapshotDone events are processed
+	requestsUpdated := func() bool {
+		requests, err := l.PendingSnapshotRequests()
+		require.NoError(t, err)
+		return equal(requests, []uint64{15, 20, 50, 100})
+	}
+	require.Eventually(t, requestsUpdated, time.Minute, time.Second)
+
+	// Test 4: commit blocks, verify snapshots are generated and pending snapshot requests are updated
+	testutilCommitBlocks(t, l, bg, 20, protoutil.BlockHeaderHash(lastBlock.Header))
+	// wait until snapshots for height=15 and height=20 are generated
+	snapshotExists = func() bool {
+		exists, err := kvledger.snapshotExists(15)
+		require.NoError(t, err)
+		return exists
+	}
+	require.Eventually(t, snapshotExists, time.Minute, 3*time.Second)
+	snapshotExists = func() bool {
+		exists, err := kvledger.snapshotExists(20)
+		require.NoError(t, err)
+		return exists
+	}
+	require.Eventually(t, snapshotExists, time.Minute, 3*time.Second)
+
+	// wait until previous snapshotDone events are processed
+	requestsUpdated = func() bool {
+		requests, err := l.PendingSnapshotRequests()
+		require.NoError(t, err)
+		return equal(requests, []uint64{50, 100})
+	}
+	require.Eventually(t, requestsUpdated, time.Minute, time.Second)
+
+	// Test 5: reopen the provider and ledger and verify snapshot requests are correct
+	provider.Close()
```

Comment:
A test for recovery code can be added (perhaps, a separate test). For this, you can separately open the bookkeeper and add a request equal to commit height and then re-open the provider.